Add redact helper script

DJensen94 opened the pull request:
Fixes #118 
## 🗣 Description ##
Add a script that can redact pii from user specified dataframe columns

## 💭 Motivation and context ##
Cyber Six gill pulls in a lot of data from the darkweb that may contain personally identifiable information that we do not want to be liable for storing. This script identifies and redacts this information before it is inserted into the database

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

Created a pandas dataframe filled with strings containing each of the different types PII we hope to redact.  After running the dataframe through this script I compared the original and the redacted version to verify that each occurrence of PII had been successfully redacted.

## ✅ Pre-approval checklist ##

- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [X] All relevant type-of-change labels have been added.
- [X] All new and existing tests pass.

DJensen94 committed:
Add redact helper script

Add a script that can redact pii from dataframe columns

DJensen94 committed:
Move script to helpers folder

Move the file to a more accessible folder

DJensen94 committed:
Add NLP redact to script

removed uneccessary comments and reincorporated NLP redaction into the code

DJensen94 committed:
add back lambda

add back lambda

DJensen94 committed:
Add second lambda back in

Add lambda for NLP redaction

DJensen94 committed:
Merge branch 'develop' into DJ_add_redact_helper

DJensen94 commented:
@dav3r This is the next PR ready for review

DJensen94 committed:
Merge branch 'DJ_add_redact_helper' of https://github.com/cisagov/pe-reports into DJ_add_redact_helper

DJensen94 commented:
@dav3r, actually pushing some commits that I forgot to push, so will verify once tests all pass

DJensen94 commented:
@dav3r ok all ready now

cduhn17 reviewed (approved):
Looks good to me.

dav3r commented:
@DJensen94 - Can you please fill out the "Testing" section of this PR description to explain how you tested this?

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+import scrubadub
+import scrubadub.detectors.date_of_birth
+
+FL = [
```
Can you please add a comment here to explain what these regexes represent?

DJensen94 replied:
@dav3r issue is closed in the following commit:[c0a9bf0](https://github.com/cisagov/pe-reports/pull/190/commits/c0a9bf02d293c40431b4782e36eaac53c883912e)

DJensen94 replied:
@dav3r issue resolved in the following commit: [c0a9bf0](https://github.com/cisagov/pe-reports/pull/190/commits/c0a9bf02d293c40431b4782e36eaac53c883912e)

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+
+
+class HI_NE_VA_DLFilth(scrubadub.filth.Filth):
+    """Create filth class for HI and NE drivers licenses."""
```
```suggestion
    """Create filth class for HI, NE, and VA drivers licenses."""
```

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+
+
+class HI_NE_VA_DLDetector(scrubadub.detectors.RegexDetector):
+    """Create detector class to identify HI, NE and VA drivers licenses."""
```
Oxford comma for the win! 😃 
```suggestion
    """Create detector class to identify HI, NE, and VA drivers licenses."""
```

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+
+
+class MN_FL_MD_MI_DLDetector(scrubadub.detectors.RegexDetector):
+    """Create detector class to identify MN, FL, MD and MI drivers licenses."""
```
More Oxford commas:
```suggestion
    """Create filth class for MN, FL, MD, and MI drivers licenses."""

    type = "MN_FL_MD_MI_drivers_license"


class MN_FL_MD_MI_DLDetector(scrubadub.detectors.RegexDetector):
    """Create detector class to identify MN, FL, MD, and MI drivers licenses."""
```

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+# Build detectors to find Drivers License ID
+
+
+class FL_DLFilth(scrubadub.filth.Filth):
```
Can you please sort these classes alphabetically by name?

DJensen94 replied:
@dav3r issue is closed in the following commit: [4119fd7](https://github.com/cisagov/pe-reports/pull/190/commits/4119fd78f0c4dd56d6fbf89fc40fc23008104a51)

DJensen94 replied:
@dav3r issue resolved in the following commit: [4119fd7](https://github.com/cisagov/pe-reports/pull/190/commits/4119fd78f0c4dd56d6fbf89fc40fc23008104a51)

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+class CA_DLFilth(scrubadub.filth.Filth):
+    """Create filth class for CA drivers licenses."""
+
+    type = "ca_drivers_license"
```
Is there a reason why this isn't
```suggestion
    type = "CA_drivers_license"
```
to match the others above it?

Similar changes should be made below.

DJensen94 replied:
@dav3r the rest of these issues are addressed in the following commit: [a0bde91](https://github.com/cisagov/pe-reports/pull/190/commits/a0bde91cdf6860f0014d3ae8d869bafba112d414)

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+
+    name = "IO_drivers_license"
+    regex = re.compile("|".join(IO), re.IGNORECASE)
+    filth_cls = IO_DLFilth
```
There is no state/territory abbreviated as `IO`.  https://about.usps.com/who/profile/history/state-abbreviations.htm

DJensen94 replied:
@dav3r resolved in the following commit
[1af3965](https://github.com/cisagov/pe-reports/pull/190/commits/1af39651dea6ea311b920c3bd85f421f2757cf89)

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+
+
+class SSNDetector(scrubadub.detectors.RegexDetector):
+    """Create detector class to identify social security numbers."""
```
```suggestion
    """Create detector class to identify Social Security numbers."""
```

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+
+    name = "no_space_ssn"
+    regex = re.compile(
+        r"(?:(?<=\s)|(?<=^))(social security number|Social Security No|Social Security #|social security number|social|Social|ssn)\W*(?!219099999|078051120)(?!666|000|9\d{2})\d{3}(?!00)\d{2}(?!0{4})\d{4}(?=$|\s)",
```
Remove duplicated clauses:
```suggestion
        r"(?:(?<=\s)|(?<=^))(social security number|Social Security No|Social Security #|social|ssn)\W*(?!219099999|078051120)(?!666|000|9\d{2})\d{3}(?!00)\d{2}(?!0{4})\d{4}(?=$|\s)",
```
(`Social` is redundant because you already have `social` and you are ignoring case.)

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+    filth_cls = PassportFilth
+
+
+# Build a detector that identifies Alien Id numbers
+class AlienIdFilth(scrubadub.filth.Filth):
+    """Create filth class for alien identification numbers."""
+
+    type = "alien id"
+
+
+class AlienIdDetector(scrubadub.detectors.RegexDetector):
+    """Create detector class to identify alien identification numbers."""
```
Looks like the official term for these are [Alien Registration Numbers](https://www.uscis.gov/tools/glossary).
```suggestion
# Build a detector that identifies Alien Registration Numbers
class AlienIdFilth(scrubadub.filth.Filth):
    """Create filth class for Alien Registration Numbers."""

    type = "alien id"


class AlienIdDetector(scrubadub.detectors.RegexDetector):
    """Create detector class to identify Alien Registration Numbers."""
```

I'll leave it up to you to rename the class names  and modify the `"alien id"` strings appropriately.

DJensen94 replied:
@dav3r issue is closed in the following commit: [6e10627](https://github.com/cisagov/pe-reports/pull/190/commits/6e106273ced7e5d9d93bd4ddbb3762b2de5f5391)

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+    filth_cls = AlienIdFilth
+
+
+# Createe various regex identifiers
```
```suggestion
# Create various regex identifiers
```

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+    analyzer = AnalyzerEngine()
+    anonymizer = AnonymizerEngine()
+    entities = [
+        "PHONE_NUMBER",
+        "CREDIT_CARD",
+        "US_DRIVER_LICENSE",
+        "US_SSN",
+        "EMAIL_ADDRESS",
+        "IP_ADDRESS",
```
Can this list be alphabetized without breaking functionality?
```suggestion
        "CREDIT_CARD",
        "EMAIL_ADDRESS",
        "IP_ADDRESS",
        "PHONE_NUMBER",
        "US_DRIVER_LICENSE",
        "US_SSN",
```

dav3r reviewed (changes requested):
Please consider my feedback and suggestions and let me know if you have any questions..

DJensen94 committed:
Alphabetize list of redaction subjects

Alphabetize the list of items to look for in the redaction

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix spelling issue

Fix spelling error in comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
add correct commas and naming for class

Add correct naming and spacing for class comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Update naming and commas of class comment

Update naming and commas of class comment to show all states involved

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add oxford comma to comments

Add oxford comma to class definition comments

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Capitalize Social Security in comment

Capitalize the words "Social Security" in the comment for the detector class

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Use correct terminology for Alien Registration Numbers

Change instances of "Alien Identification Number" to "Alien Registration Number"

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove redundant case variations.

Remove repeated clauses with different cases

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Capitalize state abbreviation

Capitalize the California State abbreviation to match other type fields

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix Indiana's abbreviation

Replace all "IO" with "IN" Indiana's correct abbreviation

DJensen94 committed:
Add comment describing what the regexes represent

Explain that the list of regexes is used to identify each states drivers license format

DJensen94 committed:
Alphabetize the classes and regexes

Update the order of the regexes and class definitions to be in alphabetical order

DJensen94 committed:
Change "alien id" to "alien registration"

Replace references of "id" with "registration number" for the alien registration classes

DJensen94 committed:
Capitalize all state abbreviations

Capitalize all instances of state abbreviations throughout the script

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+KY = [r"(?:(?<=\s)|(?<=^))[a-zA-Z]\d{2}-\d{3}-\d{3}(?=$|\s)"]
+MD = [r"(?:(?<=\s)|(?<=^))[a-zA-Z]-\d{3}-\d{3}-\d{3}-\d{3}(?=$|\s)"]
+MI = [r"(?:(?<=\s)|(?<=^))[a-zA-Z] \d{3} \d{3} \d{3} \d{3}(?=$|\s)"]
+MN_FL_MD_MI = [r"(?:(?<=\s)|(?<=^))[a-zA-Z]\d{12}(?=$|\s)"]
```
Since FL, MD, and MI already have their own regexes, I think it's cleaner to include this regex in those individual state regexes and rename this one from `MN_FL_MD_MI` to just `MN`.  I suppose that's a matter of preference, so feel free to disagree with me here.

DJensen94 replied:
@dav3r I think we want to keep the combined regex, because the redaction won't be able to tell which state these same patterns come from if we put it in each state and we want to be able to tell the end user that this specific redaction could be any of those states

dav3r replied:
Understood- can you please add a comment to that effect somewhere nearby to reduce any future confusion about this?

DJensen94 replied:
@dav3r issue resolved in the following commit: [2ccd8e8](https://github.com/cisagov/pe-reports/pull/190/commits/2ccd8e832633b9d735dabf21daff9be9299da256)

dav3r replied:
That looks great! 👍

dav3r commented on src/pe_source/data/helpers/redact_pii.py:
```diff
@@ -0,0 +1,560 @@
+    r"(?:(?<=\s)|(?<=^))[a-zA-Z]\d{3}-\d{3}-\d{2}-\d{3}-\d(?=$|\s)",
+    r"(?:(?<=\s)|(?<=^))[a-zA-Z]-\d{3}-\d{3}-\d{3}-\d{3}(?=$|\s)",
+]
+HI_NE_VA = [r"(?:(?<=\s)|(?<=^))[a-zA-Z]\d{8}(?=$|\s)"]
```
Same note as below about moving the `VA` regex out of here and into its individual regex below.

DJensen94 replied:
@dav3r see response below

DJensen94 committed:
Updated regex comments to explain combo regexes

Add comments to explain why their are regexes that link to multiple states

DJensen94 committed:
Add space after regex list

Add new line to improve readability

dav3r reviewed (approved):
Nice work here- thanks (as always) for considering my feedback. 🦅
